snuba/clickhouse/optimize/optimize.py

```diff
@@ -286,42 +327,74 @@ def optimize_partition_runner(
             f"Running schedule with cutoff time: "
-            f"{schedule.cutoff_time} with {num_threads} threads"
+            f"{schedule.cutoff_time} with {configured_num_threads} threads"
         )
-        threads: MutableSequence[threading.Thread] = []
-        for i in range(0, num_threads):
-            threads.append(
-                threading.Thread(
-                    target=optimize_partitions,
-                    args=(
+
+        with ThreadPoolExecutor(max_workers=32) as executor:
+            futures = []
+            for _ in range(min(len(partitions_to_optimize), configured_num_threads)):
+                scheduled_partition_and_jitter_pair = partitions_to_optimize.popleft()
+                futures.append(
+                    executor.submit(
+                        optimize_partitions,
                         clickhouse,
                         database,
                         table,
-                        schedule.partitions[i],
+                        [scheduled_partition_and_jitter_pair.partition],
                         schedule.cutoff_time,
                         tracker,
                         clickhouse_host,
-                        schedule.start_time_jitter_minutes[i]
-                        if schedule.start_time_jitter_minutes
-                        else None,
-                    ),
+                        scheduled_partition_and_jitter_pair.start_time_jitter_minutes,
+                    )
                 )
-            )
 
-            threads[i].start()
+            while partitions_to_optimize:
+                _, active_futures = concurrent.futures.wait(
+                    futures, return_when=concurrent.futures.FIRST_COMPLETED
+                )
 
-        # Wait for all threads to finish. They would finish either because all
-        # work is done or because a cutoff time was reached. We won't know the
-        # reason why the threads finished.
-        for i in range(0, num_threads):
-            threads[i].join()
+                configured_num_threads = typing.cast(
+                    int, get_config(_OPTIMIZE_PARALLEL_THREADS_KEY, default_num_threads)
+                )
+                for _ in range(configured_num_threads - len(active_futures)):
+                    if not partitions_to_optimize:
+                        break
+
+                    scheduled_partition_and_jitter_pair = (
+                        partitions_to_optimize.popleft()
+                    )
+                    futures.append(
+                        executor.submit(
+                            optimize_partitions,
+                            clickhouse,
+                            database,
+                            table,
+                            [scheduled_partition_and_jitter_pair.partition],
+                            schedule.cutoff_time,
+                            tracker,
+                            clickhouse_host,
+                            None,
```

Comment:
@mcannizz This part of the code adds in futures that weren't part of the first `configured_num_threads` optimize jobs. I think by passing in `None` for `start_jitter`, we effectively get rid of jittering

Reply:
Ah, I see that now. Two things:
1. I don't think we can reasonably expect a reader of the code to figure out that intention as it is right now. The difference between the calls to `optimize_partitions` is quite easy to miss.
2. I think there's a simpler way to implement this that likely removes the need for `scheduled_partition_and_jitter_pairs` entirely.

On my second point, `optimize_partition_runner` is supposed to be responsible for scheduling and coordinating the work while `optimize_partitions` is supposed to actually optimize partitions. Jitter is a scheduling concern, so we should seek to implement it entirely in `optimize_partition_runner` if we can.

How can we do that? We could try something like this:
```
def jitter_start_time(start_jitter):
    if start_jitter:    
        logger.info(
            f"{threading.current_thread().name}: Jittering start time by"
            f" {start_jitter} minutes"
        )
        time.sleep(start_jitter * 60)
```

And then:
```
# Ramp up load on ClickHouse slowly, over OPTIMIZE_PARALLEL_MAX_JITTER_MINUTES
# by forcing each thread other than the first to wait before starting.
for jitter in schedule.start_time_jitter():
    futures.append(executor.submit(jitter_start_time, jitter)

while partitions_to_optimize:
    _, active_futures = concurrent.futures.wait(
        futures, return_when=concurrent.futures.FIRST_COMPLETED
    )

    # add futures that call optimize_partitions as before
    # ...
```

What do you think about that idea? Happy to discuss sync if you'd like.